### PR TITLE
throw an exception if required filter is not set

### PR DESCRIPTION
**Pull request by jdeniau**
| Q             | A
| ------------- | ---
| Bug fix?      | kind of
| New feature?  | yes
| BC breaks?    | if filter config was wrongly set to `required=true`
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1627
| License       | MIT
| Doc PR        | Doc PR will follow if PR is approved

This PR is a start of query parameter validation.

It does only check `required=true` filter for now but I will open other PR to include [the rest of OpenAPI specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#parameter-object) on field validation.

There are some points that I am not really sure of:

  * namespace of the new `QueryParameterValidateLister`: I put it in `Core\Validator\EventListener` but maybe in `Symfony\Bridge` ?
  * I set the priority of the event listener to 16, this way, it is the first event in the queue (according to [the documentation](https://api-platform.com/docs/core/events))

Otherwise, the code is quite simple and leverage Symfony validation system.

Thanks !

### CODE REVIEW

**Comment by dunglas** [comment]
I'm glad that you work on this! Thanks.

Do we really need a dependency to the Symfoyn validator for that? Can't we just use the filter extension? It will reusing API Platform in a non-Symfony context (PSR-7 and Laravel support is on our roadmap).

About the namespace, I would suggest to put it in the `Filter` namespace.

**Comment by jdeniau** [comment]
@dunglas I used Symfony validation for several reason: 

  * api-platform automatically throw a 400 exception when a `ConstraintViolationList` is throwned
  * api-platform do already depends on a `validator` service (is it deprecated ?)
  * OpenApi declare several possible requirements that seems to work really great with Symfony constraints system

Or maybe I did not really understand your comment ?

**Comment by dunglas** [comment]
The integration with the validator is 100% optional (as with Doctrine etc), in fact it's only used in the bridge for convenience for the end user, but IIRC, not for any "core" feature like this one.

Regarding the 400, it's just one line to add to support a new exception: https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php#L251

**Comment by jdeniau** [comment]
@dunglas OK for the `ContraintViolationException`, I can throw an api-platform specific exception, but don't you think it will be reinventing the wheel to re-implement every OpenApi validation ? (enum, pattern, min / max, empty value, type, etc.) ?

I think it will be a big copy/paste from symfony's constraints, don't you ?

**Comment by jdeniau** [comment]
@dunglas I remove the dependency on Symfony validator service.

About using the FilterExtension, I did not, because FilterExtension is Doctrine-specific and will not work when using resources that are not Doctrine's.

I think it's good for merge. I will make other PR's implementing the rest of OpenApi parameters specification after that.

File: src/Exception/FilterValidationException.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Exception;
+
+/**
+ * Deserialization exception.
```
Comment (dunglas) [inline src/Exception/FilterValidationException.php]: Should be updated

File: src/Exception/FilterValidationException.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        $this->constraintViolationList = $constraintViolationList;
+
+        parent::__construct($message ?: $this->__toString(), $code, $previous);
+    }
+
+    /**
+     * Gets constraint violations related to this exception.
+     *
+     * @return ConstraintViolationListInterface
```
Comment (dunglas) [inline src/Exception/FilterValidationException.php]: You can use a return type instead.

File: src/Filter/QueryParameterValidateListener.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+use ApiPlatform\Core\Api\FilterLocatorTrait;
+use ApiPlatform\Core\Exception\FilterValidationException;
+use ApiPlatform\Core\Metadata\Resource\Factory\ResourceMetadataFactoryInterface;
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Validate query parameters depending on filter description.
```
Comment (dunglas) [inline src/Filter/QueryParameterValidateListener.php]: Validates

File: src/Filter/QueryParameterValidateListener.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
+        $resourceFilters = $resourceMetadata->getCollectionOperationAttribute($operationName, 'filters', [], true);
+
+        foreach ($resourceFilters as $filterId) {
+            if (!$filter = $this->getFilter($filterId)) {
+                continue;
+            }
+
+            foreach ($filter->getDescription($attributes['resource_class']) as $name => $data) {
+                if ($data['required'] ?? false) {
```
Comment (dunglas) [inline src/Filter/QueryParameterValidateListener.php]: Can you invert this condition to reduce the complexity?

```php
if (!($data['required'] ?? false)) {
    continue;
}
```

File: src/Filter/QueryParameterValidateListener.php [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                    $errorList = [];
+
+                    if (null === $request->query->get($name)) {
+                        $errorList[] = sprintf('Query parameter `%s` is required', $name);
```
Comment (dunglas) [inline src/Filter/QueryParameterValidateListener.php]: Can you change the backticks with double quotes for consistency with other exceptions?

File: src/Filter/QueryParameterValidateListener.php [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                    }
+
+                    if (count($errorList) > 0) {
```
Comment (dunglas) [inline src/Filter/QueryParameterValidateListener.php]: Can be simplified in `if ($errorList) {`

File: src/Filter/QueryParameterValidateListener.php (hunk 4 above)
Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: I changed a litte this part to prepare the next step that will validate several parameter rules, but this should be more understandable though

**Comment by jdeniau** [comment]
@dunglas done

File: src/Filter/QueryParameterValidateListener.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        foreach ($resourceFilters as $filterId) {
+            if (!$filter = $this->getFilter($filterId)) {
+                continue;
+            }
+
+            foreach ($filter->getDescription($attributes['resource_class']) as $name => $data) {
+                $errorList = [];
+
+                if (($data['required'] ?? false) && null === $request->query->get($name)) {
```
Comment (meyerbaptiste) [inline src/Filter/QueryParameterValidateListener.php]: This condition will not work with serializer's filters (assuming they are required) for example. Indeed, `property[]` is not equal to `property`.

File: src/Filter/QueryParameterValidateListener.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Validates query parameters depending on filter description.
+ *
+ * @author Julien Deniau <julien.deniau@gmail.com>
+ */
+class QueryParameterValidateListener
```
Comment (meyerbaptiste) [inline src/Filter/QueryParameterValidateListener.php]: Can you add a unit test please?

Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: Is it usefull ? It's fully tested with behat tests.

File: src/Filter/QueryParameterValidateListener.php (hunk 7 above)
Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: Sorry but I did not really understand what you mean here ?

Can you provide an example when this does not work ?

Comment (meyerbaptiste) [reply src/Filter/QueryParameterValidateListener.php]: Let's take the description of the property filter: https://github.com/api-platform/core/blob/63d555ae0020dc6c66feef86ff56ac5131b1f7b3/src/Serializer/Filter/PropertyFilter.php#L68-L77
And let's say `required` is equal to `true` for our example.

So, the `$name` var will be equal to `property[]` here. But with `?property['relatedDummy'][]=name` the name will be `property` in the parameter bag and `$request->query->get('property[]')` will return `null`...

File: src/Filter/QueryParameterValidateListener.php (hunk 8 above)
Comment (meyerbaptiste) [reply src/Filter/QueryParameterValidateListener.php]: Yes, Behat is only for integration and acceptance tests. You also have to unit test your class.

Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: Done

**Comment by jdeniau** [comment]
@meyerbaptiste Done for the unit tests and the "array notation" (but I'm not really happy with the implementation though :| )

File: src/Filter/QueryParameterValidateListener.php [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            }
+
+            foreach ($filter->getDescription($attributes['resource_class']) as $name => $data) {
+                $errorList = [];
+
+                if (!($data['required'] ?? false)) { // property is not required
+                    continue;
+                }
+
+                if (false !== strpos($name, '[')) { // array notation of filter
```
Comment (meyerbaptiste) [inline src/Filter/QueryParameterValidateListener.php]: I have another dirty solution:
```php
parse_str($name, $query);
$name = array_keys($query)[0] ?? '';
```
It works with:
- `'foo'`
- `'foo[]'`
- `'foo[bar]'`
- `'foo[bar][]'`
- `'foo[bar][baz]'`
- etc.

😄

Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: @meyerbaptiste done

File: src/Filter/QueryParameterValidateListener.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    /**
+     * Test if required filter is valid. It validates array notation too like "required[bar]"
+     */
+    private function isRequiredFilterValid($name, $request): bool
+    {
+        $matches = [];
+        if (false === mb_parse_str($name, $matches)) {
```
Comment (meyerbaptiste) [inline src/Filter/QueryParameterValidateListener.php]: Huum, this require the `mbstring` extension...

Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: Oh yes, I though it was included since PHP7, my bad, I remove this

Comment (jdeniau) [reply src/Filter/QueryParameterValidateListener.php]: done

**Comment by jdeniau** [comment]
@meyerbaptiste changed regex → parse_str

**Comment by jdeniau** [comment]
@Simperfit done

File: tests/Filter/QueryParameterValidateListenerTest.php [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+use ApiPlatform\Core\Tests\Fixtures\TestBundle\Entity\Dummy;
+use PHPUnit\Framework\TestCase;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+class QueryParameterValidateListenerTest extends TestCase
+{
+    private $testedInstance;
+
```
Comment (soyuka) [inline tests/Filter/QueryParameterValidateListenerTest.php]: useless newline

File: src/Filter/QueryParameterValidateListener.php [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    /**
+     * Test if required filter is valid. It validates array notation too like "required[bar]".
+     */
+    private function isRequiredFilterValid($name, $request): bool
+    {
+        $matches = [];
+        parse_str($name, $matches);
+        if (empty($matches)) {
```
Comment (soyuka) [inline src/Filter/QueryParameterValidateListener.php]: `!$matches`

File: src/Filter/QueryParameterValidateListener.php [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            return false;
+        }
+
+        $rootName = array_keys($matches)[0] ?? '';
+        if (!$rootName) {
+            return false;
+        }
+
+        if (is_array($matches[$rootName])) {
```
Comment (soyuka) [inline src/Filter/QueryParameterValidateListener.php]: `\is_array`

File: src/Filter/QueryParameterValidateListener.php [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+            $keyName = array_keys($matches[$rootName])[0];
+
+            $queryParameter = $request->query->get($rootName);
+
+            return is_array($queryParameter) && isset($queryParameter[$keyName]);
```
Comment (soyuka) [inline src/Filter/QueryParameterValidateListener.php]: `\is_array`

File: src/Filter/QueryParameterValidateListener.php [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+                    continue;
+                }
+
+                if (!$this->isRequiredFilterValid($name, $request)) {
+                    $errorList[] = sprintf('Query parameter "%s" is required', $name);
+                }
+
+                if ($errorList) {
+                    throw new FilterValidationException($errorList);
+                }
```
Comment (soyuka) [inline src/Filter/QueryParameterValidateListener.php]: shouldn't we throw outside of the foreach? if not why is `$errorList` an array?

**Comment by jdeniau** [comment]
@soyuka done.

About the `$errorList`, I changed it, but it was changed in #1723 anyway

**Comment by jdeniau** [comment]
@soyuka @meyerbaptiste @dunglas Can you please review this PR ? Every comments you made have been treated.

After that, I would really appreciate your comment about the PR following this one https://github.com/api-platform/core/pull/1723#issuecomment-375695568

Thanks

File: src/Exception/FilterValidationException.php [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Exception;
+
+/**
+ * Filter validation exception.
+ *
+ * @author Julien DENIAU <julien.deniau@gmail.com>
+ */
+class FilterValidationException extends \Exception implements ExceptionInterface
```
Comment (dunglas) [inline src/Exception/FilterValidationException.php]: `final`

File: src/Filter/QueryParameterValidateListener.php (hunk 8 above)
Comment (dunglas) [reply src/Filter/QueryParameterValidateListener.php]: `final`

File: src/Filter/QueryParameterValidateListener.php [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Filter;
```
Comment (dunglas) [inline src/Filter/QueryParameterValidateListener.php]: Please move it in the `ApiPlatform\EventListener` namespace.
What about `ValidateQueryParametersListener` for the name?

File: src/Filter/QueryParameterValidateListener.php [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+use ApiPlatform\Core\Util\RequestAttributesExtractor;
+use Psr\Container\ContainerInterface;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Validates query parameters depending on filter description.
+ *
+ * @author Julien Deniau <julien.deniau@gmail.com>
+ */
+class QueryParameterValidateListener
```
Comment (soyuka) [inline src/Filter/QueryParameterValidateListener.php]: `final`?

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php [hunk 19]
```diff
@@ -516,6 +521,7 @@ private function getPartialContainerBuilderProphecy($test = false)
             'api_platform.listener.view.respond',
             'api_platform.listener.view.serialize',
             'api_platform.listener.view.validate',
+            'ApiPlatform\Core\Filter\QueryParameterValidateListener',
```
Comment (dunglas) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php]: Please change to `api_platform.listener.view.validate_query_parameters` for consistency (we don't use id as services name yet)

File: tests/Fixtures/TestBundle/Entity/ArrayFilterValidator.php [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    public function setId($id)
+    {
+        $this->id = $id;
+    }
```
Comment (soyuka) [inline tests/Fixtures/TestBundle/Entity/ArrayFilterValidator.php]: For simplicity's sake you could have put public properties here, that's find as-is anyway! 
Although, are you sure you want `id` to be writable?

File: tests/Fixtures/TestBundle/Entity/FilterValidator.php [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    public function setId($id)
+    {
+        $this->id = $id;
+    }
```
Comment (soyuka) [inline tests/Fixtures/TestBundle/Entity/FilterValidator.php]: same comment id is auto generated

**Comment by dunglas** [comment]
About the general design, wouldn't it be better to validate the "query parameters" directly in the filter instead of in a listener? It would allow to validate required filters in GraphQL too. It would also make it easier to support and maintain this feature when we'll add an integration with Laravel, or PSR-7 support.
Now that the infrastructure has been created to access to query parameters in filter through the context, it will not be very difficult. You may move the validation logic in a trait, and use this trait in every filter to validate it.

File: tests/Fixtures/TestBundle/Filter/ArrayRequiredFilter.php [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Tests\Fixtures\TestBundle\Filter;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\AbstractFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+
+class ArrayRequiredFilter extends AbstractFilter
```
Comment (soyuka) [inline tests/Fixtures/TestBundle/Filter/ArrayRequiredFilter.php]: could be final

File: tests/Fixtures/TestBundle/Filter/RequiredFilter.php [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Tests\Fixtures\TestBundle\Filter;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Filter\AbstractFilter;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+
+class RequiredFilter extends AbstractFilter
```
Comment (soyuka) [inline tests/Fixtures/TestBundle/Filter/RequiredFilter.php]: could be final

File: tests/Fixtures/app/config/config_test.yml [hunk 24]
```diff
@@ -161,6 +162,14 @@ services:
         parent:    'api_platform.serializer.property_filter'
         tags:      [ { name: 'api_platform.filter', id: 'my_dummy.property' } ]
 
+    ApiPlatform\Core\Tests\Fixtures\TestBundle\Filter\RequiredFilter:
+        arguments: [ '@doctrine' ]
+        tags: [ 'api_platform.filter' ]
+
+    ApiPlatform\Core\Tests\Fixtures\TestBundle\Filter\ArrayRequiredFilter:
+        arguments: [ '@doctrine' ]
+        tags: [ 'api_platform.filter' ]
```
Comment (soyuka) [inline tests/Fixtures/app/config/config_test.yml]: we're not using spaces after `[` IIRC

File: src/Filter/QueryParameterValidateListener.php [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    /**
+     * Test if required filter is valid. It validates array notation too like "required[bar]".
+     */
+    private function isRequiredFilterValid($name, $request): bool
+    {
+        $matches = [];
+        parse_str($name, $matches);
```
Comment (teohhanhui) [inline src/Filter/QueryParameterValidateListener.php]: I've mentioned this a long time ago, but we really should stop using bad built-in PHP functions like `parse_str`. We should seriously think about using a proper URI parsing library instead: https://uri.thephpleague.com/5.0/

**Comment by teohhanhui** [comment]
@dunglas I think it's better if we avoid giving the filter implementations the responsibility of checking required filters. It's not their job and it doesn't need to be. But I agree with extracting all logic into a separate class, so as to keep the event listeners really dumb.

**Comment by dunglas** [comment]
Introducing a new service and using it in the listener looks good to me.

**Comment by jdeniau** [comment]
@soyuka about the "light" listener, the separation between the listener and the validation has been made in #1723 ( https://github.com/api-platform/core/pull/1723/files#diff-91ffc72a7db30ec816871e7141558def exactly ) : I prefer keeping code simple first, and re-factor it when needed.
With this modification, I can extract a FilterListValidator, but the only code it will contain is a `foreach($validatorList as $validator) { doValidate(); }`

@dunglas About the namespace, we talked about it previously, but I can move the listener in EventListener and move back the Filter validator in Filter namespace.

Can we keep the listener "as-is" for this PR, and make the refactoring in #1723, as the purpose of this PR is like a POC before #1723 ? (it's quite painful to rebase each time, and I think it will be more painful this time as the code is several months old).

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php (hunk 19 above)
Comment (soyuka) [reply tests/Bridge/Symfony/Bundle/DependencyInjection/ApiPlatformExtensionTest.php]: Still, you could alias it if you think that it'd be used by developers.

File: src/Filter/QueryParameterValidateListener.php (hunk 25 above)
Comment (dunglas) [reply src/Filter/QueryParameterValidateListener.php]: Is this really necessary for this use case? Maybe can we do that at some point in another PR and replace all calls to `parse_str` in 1 time.

**Comment by dunglas** [comment]
LGTM.

File: tests/Fixtures/TestBundle/Entity/ArrayFilterValidator.php (hunk 20 above)
Comment (soyuka) [reply tests/Fixtures/TestBundle/Entity/ArrayFilterValidator.php]: Sure thing! Just remove the id setter if it's auto generated though :)

**Comment by jdeniau** [comment]
@soyuka done

File: src/Filter/QueryParameterValidateListener.php (hunk 17 above)
Comment (soyuka) [reply src/Filter/QueryParameterValidateListener.php]: It's okay if it's refactored later as said in https://github.com/api-platform/core/pull/1692#issuecomment-391711723 I guess?

**Comment by jdeniau** [comment]
@dunglas @meyerbaptiste @Simperfit @soyuka everything is OK on my part I think. Waiting for a merge on master to work on the second PR that will check all documented fields

**Comment by soyuka** [comment]
Thanks @jdeniau !

**Pull request merged**